x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -352,74 +357,71 @@ public String toString() {
 
     private void prewarmCache() {
         if (prewarmCache) {
-            final List<BlobStoreIndexShardSnapshot.FileInfo> cacheFiles = snapshot().indexFiles()
+            final BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> queue = snapshot().indexFiles()
                 .stream()
                 .filter(file -> file.metadata().hashEqualsContents() == false)
                 .filter(file -> isExcludedFromCache(file.physicalName()) == false)
-                .collect(Collectors.toList());
+                .collect(Collectors.toCollection(LinkedBlockingQueue::new));
 
             final Executor executor = threadPool.executor(SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME);
-            logger.debug("{} warming shard cache for [{}] files", shardId, cacheFiles.size());
+            logger.debug("{} warming shard cache for [{}] files", shardId, queue.size());
 
-            for (BlobStoreIndexShardSnapshot.FileInfo cacheFile : cacheFiles) {
-                final String fileName = cacheFile.physicalName();
-                try {
-                    final IndexInput input = openInput(fileName, CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
+            // Start as many workers as fit into the searchable snapshot pool at once at the most
+            final int workers = Math.min(threadPool.info(SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME).getMax(), queue.size());
+            for (int i = 0; i < workers; ++i) {
+                prewarmNextFile(executor, queue);
+            }
+        }
+    }
+
+    private void prewarmNextFile(final Executor executor, final BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> queue) {
+        executor.execute(new AbstractRunnable() {
+            @Override
+            protected void doRun() throws Exception {
+                final BlobStoreIndexShardSnapshot.FileInfo file = queue.poll(0L, TimeUnit.MILLISECONDS);
```

Comment:
NIT: 

I guess technically we could 

```java
final BlobStoreIndexShardSnapshot.FileInfo file = queue.poll(0L, TimeUnit.MILLISECONDS);
if (file == null) {
    return;
}
```

before branching off to the executor at the beginning of this method. That way we save at least one pointless task on the scheduler for each shard (and an indent level in the abstract runnable :)).

Reply:
Sure, I pushed dc6d484930a61310a64020fb44dbe912c65d72fc

Reply:
Sorry I should've added another line here to the snippet, I was thinking about this:

```java
final BlobStoreIndexShardSnapshot.FileInfo file = queue.poll(0L, TimeUnit.MILLISECONDS);
if (file == null) {
    return;
}
executor.execute(new AbstractRunnable() {
```

i.e. doing the polling from the queue on the calling thread. This way:
1. We don't have to create as many tasks as we have threads to begin with. (e.g. if we're only warming 5 files, we're still creating 32 threads on the pool of which 27 will just not do anything and just poll a `null`)
2. We don't waste another task at the end of each file as well. Currently if 32 threads had work at some point, each of them will add one needless task that only polls a `null` at the end of the warming of a shard.

Hope that makes sense? :)

Reply:
My bad, I've been too quick reading your comment. I pushed 1bffb66766ad3dc53548c6c759c854ce4ffcee17